pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,359 @@
+package statefetcher
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"strconv"
+	"sync"
+	"sync/atomic"
+
+	"github.com/nspcc-dev/neo-go/pkg/config"
+	"github.com/nspcc-dev/neo-go/pkg/core/statesync"
+	gio "github.com/nspcc-dev/neo-go/pkg/io"
+	"github.com/nspcc-dev/neo-go/pkg/services/helpers/neofs"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neofs-sdk-go/client"
+	"github.com/nspcc-dev/neofs-sdk-go/container"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+// Ledger is the interface for statefetcher.
+type Ledger interface {
+	GetConfig() config.Blockchain
+	HeaderHeight() uint32
+	AddContractStorageData(kv []statesync.KeyValue, syncHeight uint32, expectedRoot util.Uint256) error
+	GetLastStoredKey() []byte
+}
+
+// Service fetches contract storage state from NeoFS.
+type Service struct {
+	neofs.BasicService
+
+	isActive   atomic.Bool
+	isShutdown atomic.Bool
+	lock       sync.RWMutex
+
+	cfg               config.NeoFSStateFetcher
+	stateSyncInterval uint32
+	lastStateIndex    uint32
+	lastStateOid      oid.ID
+
+	chain Ledger
+	log   *zap.Logger
+
+	quit             chan bool
+	quitOnce         sync.Once
+	runToExiter      chan struct{}
+	exiterToShutdown chan struct{}
+
+	shutdownCallback func()
+}
+
+// New creates a new Service instance.
+func New(chain Ledger, cfg config.NeoFSStateFetcher, stateSyncInterval int, logger *zap.Logger, shutdownCallback func()) (*Service, error) {
+	if !cfg.Enabled {
+		return &Service{}, nil
+	}
+	if cfg.Timeout <= 0 {
+		cfg.Timeout = neofs.DefaultTimeout
+	}
+	if cfg.StateAttribute == "" {
+		cfg.StateAttribute = neofs.DefaultStateAttribute
+	}
+	if cfg.KeyValueBatchSize <= 0 {
+		cfg.KeyValueBatchSize = neofs.DefaultKVBatchSize
+	}
+
+	basic, err := neofs.NewBasicService(cfg.NeoFSService)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create service: %w", err)
+	}
+
+	sfs := &Service{
+		BasicService:      basic,
+		log:               logger,
+		cfg:               cfg,
+		chain:             chain,
+		shutdownCallback:  shutdownCallback,
+		stateSyncInterval: uint32(stateSyncInterval),
+		quit:              make(chan bool),
+		runToExiter:       make(chan struct{}),
+		exiterToShutdown:  make(chan struct{}),
+	}
+
+	if sfs.stateSyncInterval == 0 {
+		sfs.stateSyncInterval = config.DefaultStateSyncInterval
+	}
+	var (
+		containerID  cid.ID
+		containerObj container.Container
+	)
+	sfs.Ctx, sfs.CtxCancel = context.WithCancel(context.Background())
+	if err = sfs.Pool.Dial(context.Background()); err != nil {
+		sfs.isShutdown.Store(true)
```

Comment:
BlockFetcher uses the same `isActive` and `isShutdown` atomic values to track service state. However, the application of these values is different, for example in the situation when pool dial failed, BlockFetcher uses:
https://github.com/nspcc-dev/neo-go/blob/44bd1936b9dc9baac8b42769be98546ae162228c/pkg/services/blockfetcher/blockfetcher.go#L207-L208

We need to unify the behaviour, in all places of StateFetcher service.